### ISSUE TITLE
Asset in /static is duplicated in /build (with/without hash)

### ISSUE THREAD

**Issue by pranabdas** [facebook/docusaurus#5362]
## 💥 Proposal

Docusaurus transfers all the files under `static` directory to root build directory. Additionally, images, PDFs linked using markdown syntax are processed (files names are suffixed with a hash) and placed under `images`, `files` etc. and these are actually linked. So there are two copies of those files. I guess this is not ideal, and probably there should be better build rules to avoid these duplicates.

However, the above is not true for every file, for example, the `favicon` in the `doucsaurus.config.js` is not given hashed filename, neither the files linked using JSX baseUrl. In this case, the website links them to the ones transferred from the `static` directory. 

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

**Comment by slorber** [comment]
Some rules we don't want to break:
- every file in `/static` will be in `/build`
- every file that is required/imported statically goes into the Webpack asset processing pipeline and is hashed.
- having a hash is better for immutable HTTP caching

> Additionally, images, PDFs linked using markdown syntax are processed (files names are suffixed with a hash) and placed under images, files etc. and these are actually linked.

We try to convert file links and images to require calls, for caching reasons. Now, nothing forces you to put those files in `/static`. If you put your files in another place, those won't be duplicated and you'd only get the hashed version (better for caching).

So, basically, you can avoid this duplication yourself today but reducing your usage of `/static`. However, I do agree that Docusaurus tend to be biased toward using `/static` (when using markdown images, we look for them in `/static`).

Maybe we should introduce a new `/assets` folder or something, that we won't copy automatically to `/build`? This means if an image in `/assets` is not required explicitly or referenced in any md doc,  it won't appear anywhere in `/builld`. Does it make sense to do that? Do you have a better name to suggest?

**Comment by pranabdas** [comment]
I see. Thank you. In my use case, `/assets` folder would make better sense. Most of my assets are part of the site i.e., they are linked/imported. For me `/static` is a special case, I would use it in case I need an absolute link, e.g., `https://example.com/file.pdf`. Is there a way to link items in `/assets` folder using absolute path something like `/static/img/profile.jpg` is linked by `/img/profile.jpg`? It would be nice if it is possible to refer them by `/assets/img/profile.jpg`, currently that path looks for `/static/assets/img/profile.jpg`.

PS: If I understood the reasoning correctly, the `/assets` should be given the current role of `/static`. I guess `/static` is only good if I need absolute links while everything linked/imported in the site are hashed and linked to the `/assets` upon build. Why would I keep website assets in a folder that would be replicated upon build but not linked to (these replicated items have no use). `/static` should be reserved for files that need to be replicated but not linked/imported in the site (e.g., `.nojekyll`).

**Comment by slorber** [comment]
> Is there a way to link items in /assets folder using absolute path something like /static/img/profile.jpg is linked by /img/profile.jpg? It would be nice if it is possible to refer them by /assets/img/profile.jpg, currently that path looks for /static/assets/img/profile.jpg.

For assets going through Webpack (`build/assets/...` with a hash), you can't just reference them with a path, you have to require them. We convert md file links and images to require calls automatically for you, but we can't do anything if you use `<img>` or `<Link>` => in this case it's your responsibility to use `require()`.

Many assets may be used on a Docusaurus website and don't necessarily go through Webpack: favicon, pwa manifest, .nojekyll, Netlify _redirects file, javadoc, static HTML files that you reference in Docusaurus iframes... 

> PS: If I understood the reasoning correctly, the /assets should be given the current role of /static.

That's not what I meant here. `/static` will keep its existing behavior otherwise it will be a breaking change that will annoy almost every Docusaurus user.

> Why would I keep website assets in a folder that would be replicated upon build but not linked to (these replicated items have no use). /static should be reserved for files that need to be replicated but not linked/imported in the site (e.g., .nojekyll).

Having assets copied in `/build/img-xyz.png` and also hashed in `/build/assets/img-xyz-md5.png` can be useful for some use-cases: you may want to leverage the caching benefits for some use-cases, but also get a predictable URL for other use-cases (both at the same time). Sometimes this duplication is actually a useful behavior.

**Comment by pranabdas** [comment]
Thank you for the details.

**Comment by slorber** [comment]
NP.

I still think having an `/assets` folder could be nice to make it easier to not have a duplicate file output in `/build`.

But I think having this duplicate output is not too bad, it can actually be useful and there are workarounds to avoid that already.

Not duplicating output files would only be useful to improve the perf of your CDN deployment time: you upload 1 file instead of 2.

Before implementing anything, I'd prefer to see if anyone has a strong use case to have this in the first place, that couldn't be solved by a workaround. 

For example, a clear and measurable perf issue in the deployment of your site due to using a lot of assets, and inability to use co-located assets for some reasons (versioning, i18n can be valid reasons)

**Comment by pranabdas** [comment]
@slorber I am going to adopt `/assets` for certain cases, e.g., a large directory of PDFs. It would be good if it gets the benefits of `/static`. Alternatively, a Jekyll-like `_dirName` (or say, `EXCLUDE_dirName`) to exclude directories/files in the `/static` during build would be useful. Thank you.

**Comment by slorber** [comment]
Can you link to a doc for Jekyll-like _dirName? I don't know anything about Jekyll sorry

**Comment by pranabdas** [comment]
@slorber By default files/directories starting with `.`, `_`, `#` or `~` in the source directory are excluded during Jekyll build https://jekyllrb.com/docs/structure/ So the convention is almost all source files/directory names to be starting with underscore, so that they are not copied as such during build rather only the Jekyll processed ones according to its config (e.g., `_blog/post.md` →`/blog/2021/08/post/index.html`), and directories (e.g., `assets`) that need to be copied are given usual names like `assets`. Additionally, they have `include` and `exclude` options in the configuration to handle exceptions.

**Comment by slorber** [comment]
Let's keep it open then, not sure we'll adopt exactly the same convention through

**Comment by dirslashls** [comment]
I ran into this same issue of two copies of the file. In my case, I have a `page` referencing a pdf file. Instead of putting the pdf in static directory I put it in `pages/assets` directory and referenced it in the page.md as `(pdf)[./assets/pdf.pdf]`. With this, it just created a file in build/assets/files/pdf-[hash].pdf

**Comment by slorber** [comment]
@dirslashls this is the expected behavior that we actually want to keep. Why don't you like it and what is your problem in practice?

**Comment by dirslashls** [comment]
This helps not have duplicate files. Only thing missing is the not having hash in the file name. I am hoping for a permalink where the most recent content of the file is always accessed with the same URL. At the moment, my solution is to provide the page url as the permalink and the user has to access the file linked on the page.

**Comment by pranabdas** [comment]
Last time I saw using `baseUrl` does not add hashes to the file names. You need to put the files under `/static`, so they are copied over to build accordingly. 
```
<!-- in the top of your markdown file -->
import useBaseUrl from '@docusaurus/useBaseUrl';

<!-- while linking -->
Download <a href={useBaseUrl("/pdf/file.pdf")}>PDF</a>. 

<img src={useBaseUrl("/img/profile.png")} alt="profile" width="400px" />
```
The above files are located at `/static/pdf/file.pdf` and `/static/img/profile.png`, respectively.

**Comment by dirslashls** [comment]
I actually tried this. The problem is that directly accessing the link (/pdf/file.pdf) worked but clicking on the link in the page generated using useBaseURL gives page not found error even though both urls are same.

**Comment by pranabdas** [comment]
I see. Actually, I had inline images and pdf in new tab `<a target="_blank" href={useBaseUrl("/pdf/file.pdf")}>PDF</a>`, both worked.

**Comment by slorber** [comment]
The rules are:
- files in `/static` will be copied as is, no hash in filename
- files not in `/static` will NOT be copied, so they will always end-up with a hash, only if those files are required in your site's js/md
- files in `/static` may end-up being both copied with/without hash (if they are required in your site's js/md): this is the issue reported here

By default, we really want to have hashes in the URL for aggressive caching and caching invalidation reasons. 
Not having hashes in the URL is not the default behavior on purpose and we don't plan to change that.

If you want stable URLs, always use `/static`

If you use `<a href="">PDF</a>` in a md doc, Docusaurus automatically converts it to `<Link href="">PDF</Link>` which will try to "navigate" to the URL you provide (using SPA navigation, `history.push()` API), because this is generally what you want to do when using a markdown link (downloading a file is way less common than navigating across md docs).

As @pranabdas explained, using `target="_blank"` will not use SPA navigation and will simply open the link in a new window: that's a good solution to download a file from `/static`

**Comment by Josh-Cena** [comment]
Going to close this as working as intended. If you don't want the asset to appear in `/assets`, you can use `pathname:///file.pdf`, which will also be rendered as an external link that opens a new tab.
